README.md
# Swedbank Pay Design Guide

[![Version][version-badge]][version-tag]
![Tag][github-actions-badge]
[![Codecov][codecov-badge]][codecov-status]
[![Dependabot Status][dependabot-badge]][dependabot]
[![Dependencies Status][dep-badge]][dep-status]
[![devDependencies Status][devdep-badge]][devdep-status]

[![Swedbank Pay Design Guide][og-image]][swpdg]

## Installing / Getting started

The following guides are for further development of the Design Guide, and are
not necessary if you only plan to use the Design Guide in your project.

### Prerequisites

Design Guide requires [Node.js][nodejs] (includes npm).

### Installing

1.  Open a terminal and navigate to repo root
2.  `npm ci`
3.  `npm start`
4.  Go to `http://localhost:3000`
5.  Works? If not, contact us on Slack ([`#design-guide-general`][slack] – only
    available to Swedbank Pay and PayEx employees).
6.  Get to work.

### Set up Git Hooks

-   Open a terminal and navigate to repo root
-   Run `git config core.hooksPath .githooks`

## Developing

It is important to note that as a developer, this project supports two brands:
**Swedbank Pay** _and_ **PayEx**. [Webpack][webpack] is configured to accept a
brand flag which is used to collect different sets of theme variables.
**Swedbank Pay** is default flag. E.g. `npm start` sets the theme to
**Swedbank Pay**, while `npm run start:payex` sets the theme to … you guessed
it, **PayEx**. Check out the `package.json` file for additional scripts.

### Built with

-   [React][react]
-   [Less][less]

### Setting up Dev

-   Running the command `npm start` starts the Webpack dev server, (which is
    configured in `~/webpack.config.js`). If any changes is made to this file, or
    any files directly influencing this file (i.e. any loaders or plugins) you
    need to restart the server.
-   The script `npm start` will start the dev server in development mode. If you
    wish to run the dev server in production-like environment run the command
    `npm run start:prod`.

#### Importing components

When importing different components in the files you create, you can use
shortcuts for common paths. The shortcuts are as following:

|    Shortcut   | Description                                                                            |
| :-----------: | :------------------------------------------------------------------------------------- |
|      `~`      | root of the project (example `import package from "~/package";`)                       |
|     `@src`    | `~/src`-folder (example `import dg from "@src/scripts";`)                              |
| `@components` | `~/src/App/components`-folder (example `import Alert from "@components/Alert";`)       |
|  `@docutils`  | `~/src/App/Documentation/utils`-folder (example `import { DocToc } from "@docutils";`) |
|  `@constants` | `~/src/constants`-folder                                                               |

The shortcuts are specified in the `~/.babelrc`-file. Specifying the
prefix-shortcuts in the `~/jsconfig.json`-file enables path intellisense for
the shortcuts (at least in VSCode).

### Building

To start building, run the command `npm run build` to start in dev-mode, or
`npm run build:prod` for production-mode. This will make webpack start bundling
everything and make static files in the `~/dist/`-folder. Also make sure all
tests pass, or make changes to them accordingly.

> Note: Running the command `npm run build:prod` will delete all files already
> located in `~/dist`.

## Versioning

Design Guide uses [semantic versioning][semver]. This means that internally,
Design Guide will use a `major.minor.patch` version number to signal what the
changes in each release surmount to. This also means both CSS and JavaScript
files are versioned in lockstep, so whenever a new version of the CSS is
released, a new version of the JavaScript will be as well.

## Tests

1.  Open a terminal and navigate to repo root
2.  Run the command `npm test`
    -   Many of the tests are snapshot based, and runs against already created
        snapshots. If you need to update the snapshots due to changes run the
        command `npm run test:update`.
    -   A test coverage report is generated after running the tests. This can be
        found in `~/coverage`.

> Note: Snapshot testing is pretty unforgiving, so if you make **ANY** changes
> to how a component, which already has a snapshot, is rendered, the tests
> **WILL** crash! So make sure to run `npm run test:update` before committing.

## Code style

Coding style rules are set in `~/.eslintrc` & `~/.stylelintrc`.

**_Make sure to follow the syntax rules enforced by ESLint & Stylelint._**

## Contributing

-   Bug reports: File bugs as Github [issues][issues].
-   Feature suggestions: File as Github issue or discuss the feature on Slack
    ([`#design-guide-general`][slack]).
-   Code contributions: Code contributions are highly encouraged, but discuss
    the feature/bugfix in an issue or on Slack before you start coding.

## Creating a release

Before merging to master to create a release, make sure you follow these steps:

1.  Create a release branch `release/x.x.x`.

2.  Make sure the linting (`npm run lint`) and unit tests pass (`npm test`).

3.  Update the version in `~/package.json`.

4.  Make sure the `RELEASE-NOTES.md` is correct (date and version) and only contains
    changes for the current release. (`CHANGELOG.md` will be updated automatically)

5.  Update the version in the URL of the `version-badge` and `version-tag`.
    at the bottom of this `README.md` file.

6.  Commit and push changes to your release branch. Two pull requests will be automatically created to `master` and `develop` branches.

7.  When approved, rename the merge commit to `Release x.x.x` and merge.

8.  Change branch to `master`.

9.  Create and push a new tag with the new release commit.

GitHub Actions will now create and deploy a release on both github and
[`design.swedbankpay.com`][swpdg].

After pushing to master, create a pull request to `develop` that deletes the
release notes content but keeps the headers.

## Core development team

-   [Erik Hallaraaker][erikhallaraaker]
-   [Håkon Mosvoll][hmosvoll]
-   [Thomas Lorntzen][vikingtristan]

## Contents of this project

| Folder       | Description                                                                    |
| :----------- | :----------------------------------------------------------------------------- |
| `~/coverage` | Test coverage report for the project (not commited to the repository).         |
| `~/dist`     | Static files generated by [Webpack][webpack] (not commited to the repository). |
| `~/src`      | All the source files.                                                          |
| `~/build`    | 404 and index files that serve as root domain redirectors to newest release.   |
| `~/tools`    | Various tools, mostly related to the build/deploy process.                     |

## Notable files

|  File               | Description                                                                                                                                            |
| :------------------ | :----------------------------------------------------------------------------------------------------------------------------------------------------- |
| `.babelrc`          | [Babel][babel] configuration.                                                                                                                          |
| `.eslintignore`     | List of files/directories ESLint will ignore (similar to `.gitignore`)                                                                                 |
| `.eslintrc`         | [Eslint][eslint] configuration.                                                                                                                        |
| `.sentryclirc`      | [Sentry][sentry] configuration used by Webpack to configure a new release on sentry.                                                                   |
| `.stylelintrc`      | [Stylelint][stylelint] configuration.                                                                                                                  |
| `CHANGELOG.md`      | Collection of changes made to the project. Insert your changes here.                                                                                   |
| `RELEASE-NOTES.md`  | Latest release notes of changes made to the project. Insert your changes here. This file is used by GitHub Actions to create a release body on GitHub. |
| `jest.config.js`    | [jest][jest] configuration.                                                                                                                            |
| `jest.setup.js`     | Script file that will run before jest executes the tests, this includes [enzyme][enzyme] setup, polyfills and other useful snippets.                   |
| `webpack.config.js` | [Webpack][webpack] configuration.                                                                                                                      |

## Copyright, license and credits

Code and documentation © Swedbank Pay and contributors, released under the [MIT License](LICENSE).

Inspired by [Bootstrap][bootstrap] and [Materialize][materialize].

[version-badge]:      https://img.shields.io/badge/Version-6.0.5-blue.svg

[version-tag]:        https://github.com/swedbankpay/design.swedbankpay.com/releases/tag/6.0.5

[github-actions-badge]: https://github.com/SwedbankPay/design.swedbankpay.com/workflows/Tag/badge.svg

[codecov-badge]: https://codecov.io/gh/swedbankpay/design.swedbankpay.com/branch/develop/graph/badge.svg

[codecov-status]: https://codecov.io/gh/swedbankpay/design.swedbankpay.com

[dependabot-badge]: https://api.dependabot.com/badges/status?host=github&repo=SwedbankPay/design.swedbankpay.com

[dependabot]: https://dependabot.com

[dep-badge]: https://david-dm.org/swedbankpay/design.swedbankpay.com/status.svg

[dep-status]: https://david-dm.org/swedbankpay/design.swedbankpay.com

[devdep-badge]: https://david-dm.org/swedbankpay/design.swedbankpay.com/dev-status.svg

[devdep-status]: https://david-dm.org/swedbankpay/design.swedbankpay.com?type=dev

[og-image]: https://repository-images.githubusercontent.com/191536903/4ef09b80-53ec-11ea-9f50-8fa24f706dfb

[swpdg]: https://design.swedbankpay.com

[nodejs]: https://nodejs.org/en/

[slack]: https://payex.slack.com/messages/C0L3W8B2S/

[react]: https://reactjs.org/

[less]: http://lesscss.org/

[webpack]: https://webpack.js.org/

[semver]: http://semver.org/

[issues]: https://github.com/swedbankpay/design.swedbankpay.com/issues

[erikhallaraaker]: https://github.com/erikhallaraaker

[hmosvoll]: https://github.com/hmosvoll

[vikingtristan]: https://github.com/VikingTristan

[babel]: https://babeljs.io/

[eslint]: https://eslint.org/

[sentry]: https://sentry.io

[stylelint]: https://stylelint.io/

[jest]: https://facebook.github.io/jest/

[enzyme]: http://airbnb.io/enzyme/

[bootstrap]: http://getbootstrap.com/

[materialize]: https://materializecss.com/


CHANGELOG.md
# Changelog

<!--- Begin Release 6.0.2 -->
## [5.0.2] - 2021-02-15 
awdawd
<!--- End Release 6.0.2 -->

## [5.0.2] - 2021-02-15 

## Summary
- Our first bugfix after last weeks major patch is now live! We made some small tweaks and patched a few minor problems that we found. No need to worry, everything is better again!

## Added
- Sidebar
  - Scrollbar on sidebar Should only appear during hover, and disappear when not

## Changed
- Colors
  - Updated success and info colors
- Typography
  - Changed color of blockquote element line, and made left padding smaller
  - Changed space between sections, h2 and h3
- Design Guide
  - Accessibility - Figure 1
  - Changelog page
    - Changed title for 5.0.0 update
    - Replaced "V" with "Version"
    - Added proper URL to Github link
    - Switched background to light brown
  - Favicon
    - Updated to a new version
  - Grid illustrations
    - Updated to a lighter color scheme to match overall design
  - Introduction - What is design
    - Update explanation paragraph to further explain the benefit of DG's
  - Logotype
    - Moved first paragraph of ingress to underneath "The black text logotype"
  - Status badges 
    - Moved status badges from under, to the right side of H1
  - Changed quotation styling

## Fixed
- Design Guide
  - Fixed cards in Identity oveview to have the right CSS styling
  - Fixed so card examples to take more space
  - Fixed black border bug in Colors 
  - Edge browser icon is to small
  - Enlarged Edge browser icon in For developers
  - Fixed documentation referring to an older version of design guide 

<!--- Begin Release 4.8.1 -->
## [4.8.1] - 14.09.2020

### Added
- Readme update
<!--- End Release 4.8.1 -->

<!--- Begin Release 4.8.0 -->
## [4.8.0] - 10.09.2020

### Added
- A bunch of hahahahaha
- Something else interesting! awda wdwad 
- awdok awpodkpawo kdopkawd dwa 

### Changed
- Something
- What

### Fixed
- Some bug
<!--- End Release 4.8.0 -->


## [4.6.0] - 2020-05-29

### Added
- awdawdawd
### Changed
- awdawdawd


.github/scripts/variables.sh
#!/bin/bash
set -o errexit # Abort if any command fails
me=$(basename "$0")

help_message="\
Generates variables based on the provided brand and ref arguments.
Usage:
    $me --brand <brand> [--ref <ref>]
    $me --help
Arguments:
    -h, --help              Displays this help screen.
    -b, --brand <brand>     The name of the brand to generate variables for.
    -r, --ref <ref>         The Git ref being built."

initialize() {
    while : ; do
        if [[ $1 = "-h" || $1 = "--help" ]]; then
            echo "$help_message"
            return 0
        elif [[ ( $1 = "-b" || $1 = "--brand" ) && -n $2 ]]; then
            brand=$2
            shift 2
        elif [[ ( $1 = "-r" || $1 = "--ref" ) && -n $2 ]]; then
            ref=$2
            shift 2
        else
            break
        fi
    done

    if [[ -z "$brand" ]]; then
        echo "No brand specified." >&2
        echo "$help_message"
        return 1
    fi
}

generate_variables() {
    if [[ "$ref" == refs/tags/* ]]; then
        version="${ref#refs/tags/}"
        echo "::set-output name=VERSION::$version"
    elif [[ "$ref" == refs/heads/release/* ]]; then
        version="${ref#refs/heads/release/}"
        echo "::set-output name=VERSION::$version"
    elif [[ "$ref" == refs/heads/* ]]; then
        branch="${ref#refs/heads/}"
        echo "::set-output name=BRANCH::$branch"
    fi

    echo "Ref:      $ref"
    echo "Branch:   $branch"
    echo "Version:  $version"
    echo "Brand:    $brand"

    if [ "$brand" == "payex" ]; then
        echo ::set-output name=BRAND_NAME::PayEx
        echo ::set-output name=AZURE_ACCOUNT_PROD::payexdesignguidetest
        echo ::set-output name=AZURE_ACCOUNT_STAGE::payexdesignguidetest
        echo ::set-output name=BUILD_SCRIPT::build:prod:payex
    elif [ "$brand" == "swedbankpay" ]; then
        echo ::set-output name=BRAND_NAME::SwedbankPay
        echo ::set-output name=AZURE_ACCOUNT_PROD::swedpaydesignguidetest
        echo ::set-output name=AZURE_ACCOUNT_STAGE::swedpaydesignguidetest
        echo ::set-output name=BUILD_SCRIPT::build:prod
    else
        echo "Unknown brand '$brand'!"
        return 1
    fi
}

main() {
    initialize "$@"
    generate_variables
}

main "$@"

.github/scripts/update-changelog.sh
#!/usr/bin/env bash
# set -o errexit # Abort if any command fails
me=$(basename "$0")
help_message="\
Usage:
  $me <version>

Arguments:
  version   The version of the current release."

version=$1

if [ -z "$version" ]; then
    echo "Missing required argument <version>."
    echo "$help_message"
    exit 1
fi

echo "Updating changelog for release ${version}"

release_notes=$(cat RELEASE-NOTES.md)
changelog=$(sed "s/^# Changelog//" CHANGELOG.md)

begin_notes="<!--- Begin Release ${version} -->"
end_notes="<!--- End Release ${version} -->"

if grep -F "Begin Release ${version}" CHANGELOG.md
then
    echo "Rewriting release ${version} notes."

    new_changelog=$(sed "1,/${end_notes}/d" CHANGELOG.md)

    echo "# Changelog

$begin_notes
$release_notes
$end_notes
$new_changelog" > CHANGELOG.md
else
    echo "Adding ${version} notes."
    echo "# Changelog

$begin_notes
$release_notes
$end_notes $changelog" > CHANGELOG.md
fi

git config user.name github-actions
git config user.email github-actions@github.com
git add .
git commit -m "Update changelog"
git status
git push


webpack.config.js
/* eslint camelcase: 0, object-curly-newline: 0 */
const path = require("path");
const webpack = require("webpack");
const appRoutes = require("./tools/generate-routes-copy-array");
const levelsToRoot = require("./tools/levels-to-root");
const autoprefixer = require("autoprefixer");
const TerserPlugin = require("terser-webpack-plugin");
const SentryCliPlugin = require("@sentry/webpack-plugin");
const HtmlWebpackPlugin = require("html-webpack-plugin");
const FileManagerPlugin = require("filemanager-webpack-plugin");
const MiniCssExtractPlugin = require("mini-css-extract-plugin");
const BundleAnalyzerPlugin = require("webpack-bundle-analyzer").BundleAnalyzerPlugin;
const OptimizeCSSAssetsPlugin = require("optimize-css-assets-webpack-plugin");
const AppManifestWebpackPlugin = require("app-manifest-webpack-plugin");

module.exports = (env, argv) => {

    const brand = argv.brand || "swedbankpay";
    const brandTitle = brand === "swedbankpay" ? "Swedbank Pay" : "PayEx"; // <-- Used with the HTML plugin for titles etc...
    const brandLink = brand === "swedbankpay" ? "https://swedbankpay.com" : "https://payex.com";
    const isProd = argv.mode === "production";
    const isDevServer = !!argv.host;
    const version = env && env.semver ? env.semver : "LOCAL_DEV";
    const isRelease = env && env.release === "true";
    const isGitHubActions = env && env.github_actions === "true";
    const basename = env && env.basename ? `/${env.basename}/` : "/";

    const config = {
        entry: {
            dg: ["@babel/polyfill", "./src/scripts/main/index.js"],
            "dg-dashboard": "./src/scripts/dashboard/index.js",
            app: ["@babel/polyfill/noConflict", `./src/${brand}.js`]
        },
        resolve: {
            extensions: [".js", ".jsx", ".json"]
        },
        output: {
            path: path.resolve(__dirname, `dist${basename}`),
            filename: "scripts/[name].js?[hash]",
            chunkFilename: "scripts/[name].js?[hash]",
            publicPath: basename
        },
        // target: "async-node",
        devtool: "source-map",
        devServer: {
            contentBase: path.resolve(__dirname, `dist${basename}`),
            publicPath: basename,
            compress: true,
            port: 3000,
            hot: true,
            clientLogLevel: "warning",
            historyApiFallback: true,
            disableHostCheck: true /* Temporary fix for the websocket issue with webpack dev server on IE11 [AW] */
        },
        module: {
            rules: [
                {
                    test: /\.jsx?$/,
                    exclude: modulePath => (
                        (/node_modules/).test(modulePath) &&
                        !(/node_modules\/*/).test(modulePath) &&
                        (/__snapshots__/).test(modulePath)
                    ),
                    use: {
                        loader: "babel-loader",
                        options: {
                            presets: ["@babel/preset-env"]
                        }
                    }
                },
                {
                    test: /\.less$/,
                    resolve: { extensions: [".less"] },
                    use: [
                        {
                            loader: isProd ? MiniCssExtractPlugin.loader : "style-loader"
                        },
                        {
                            loader: "css-loader"
                        },
                        {
                            loader: "postcss-loader",
                            options: {
                                plugins: () => autoprefixer({
                                    grid: true
                                })
                            }
                        },
                        {
                            loader: "less-loader",
                            options: {
                                javascriptEnabled: true
                            }
                        }
                    ]
                },
                {
                    test: /\.css$/,
                    use: [
                        {
                            loader: isProd ? MiniCssExtractPlugin.loader : "style-loader"
                        },
                        "css-loader",
                        {
                            loader: "postcss-loader",
                            options: {
                                plugins: () => autoprefixer()
                            }
                        }
                    ]
                },
                {
                    test: /\.(woff(2)?|ttf|eot)$/,
                    use: [
                        {
                            loader: "file-loader",
                            options: {
                                outputPath: "fonts/",
                                name: "[name].[ext]"
                            }
                        }
                    ]
                },
                {
                    test: /\.svg$/i,
                    include: [
                        path.resolve(__dirname, "src/icons/flags/1x1")
                    ],
                    use: [
                        {
                            loader: "file-loader",
                            options: {
                                outputPath: "img/flags/1x1/",
                                name: "[name].[ext]?[hash]"
                            }
                        }
                    ]
                },
                {
                    test: /\.svg$/i,
                    include: [
                        path.resolve(__dirname, "src/icons/flags/4x3")
                    ],
                    use: [
                        {
                            loader: "file-loader",
                            options: {
                                outputPath: "img/flags/4x3/",
                                name: "[name].[ext]?[hash]"
                            }
                        }
                    ]
                },
                {
                    test: /\.(png|jpe?g|gif|svg)$/i,
                    exclude: /flags/,
                    use: [
                        {
                            loader: "file-loader",
                            options: {
                                outputPath: "img/",
                                name: "[name].[ext]?[hash]"
                            }
                        }
                    ]
                }
            ]
        },
        optimization: {
            splitChunks: {
                chunks: "async",
                minSize: 30000,
                maxSize: 0,
                minChunks: 1,
                maxAsyncRequests: 5,
                maxInitialRequests: 3,
                automaticNameDelimiter: "~",
                name: true,
                cacheGroups: {
                    vendors: {
                        test: /[\\/]node_modules[\\/]/,
                        priority: -10
                    },
                    default: {
                        minChunks: 2,
                        priority: -20,
                        reuseExistingChunk: true
                    },
                    dgStyles: {
                        name: "dg-style",
                        test: brand === "swedbankpay" ? /(flatpickr\.css|swedbankpay\.less)$/ : /(flatpickr\.css|payex\.less)$/,
                        chunks: "all",
                        enforce: true
                    },
                    docStyles: {
                        name: "documentation",
                        test: brand === "swedbankpay" ? /documentation-swedbankpay\.less/ : /documentation-payex\.less/,
                        chunks: "all",
                        enforce: true
                    }
                }
            },
            minimize: isProd,
            minimizer: [
                new TerserPlugin({
                    sourceMap: true,
                    terserOptions: {
                        compress: { drop_console: true }
                    }
                }),
                new OptimizeCSSAssetsPlugin()
            ],
            mergeDuplicateChunks: isProd
        },
        plugins: [
            new HtmlWebpackPlugin({
                template: "./src/index.html",
                hash: true,
                title: `${brandTitle} DesignGuide`
            }),
            new MiniCssExtractPlugin({
                filename: "styles/[name].css"
            }),
            new webpack.DefinePlugin({
                "process.env": {
                    basename: JSON.stringify(basename),
                    version: JSON.stringify(version),
                    sentry: isRelease,
                    isGitHubActions,
                    google: isRelease,
                    brand: JSON.stringify(brand),
                    brandTitle: JSON.stringify(brandTitle)
                }
            }),
            new webpack.IgnorePlugin(/^\.\/locale$/, /moment$/) // Ignores moments locale folder which doubles the size of the package, moment is a dependency of chart.js [EH]
        ]
    };

    if (!env && !isDevServer) {
        config.plugins.push(
            new BundleAnalyzerPlugin()
        );
    }

    if (isRelease) {
        const rootPath = levelsToRoot(basename);

        config.plugins.push(
            new HtmlWebpackPlugin({
                filename: `${rootPath}index.html`,
                template: "./build/rootindex.html",
                hash: true,
                title: `${brandTitle} Design Guide`,
                chunks: ["dg"],
                basename
            }),
            new HtmlWebpackPlugin({
                filename: `${rootPath}404.html`,
                template: "./build/root404.html",
                hash: true,
                chunks: ["dg"],
                title: `${brandTitle} Design Guide`,
                basename
            }),
        );

        // Don't create new sentry release on GitHub Actions
        if (!isGitHubActions) {
            config.plugins.push(
                new SentryCliPlugin({
                    release: version,
                    include: ".",
                    ignore: ["node_modules", "webpack.config.js"]
                })
            );
        }
    }

    if (isProd && !isDevServer) {
        const onEndArchive = [
            {
                source: "./dist/temp/icons",
                destination: `./dist${basename}release/icons.zip`
            }
        ];

        if (isRelease) {
            onEndArchive.push({
                source: "./dist/temp/release",
                destination: `./dist${basename}release/${brand === "swedbankpay" ? "Swedbankpay" : "Payex"}.DesignGuide.v${version}.zip`
            });
        }

        config.plugins.push(
            new AppManifestWebpackPlugin({
                logo: `./src/img/${brand}/favicon.png`,
                prefix: basename,
                output: "/icons/",
                config: {
                    appName: `${brandTitle} DesignGuide`,
                    developerName: brandTitle,
                    developerURL: brandLink,
                    background: "#000",
                    theme_color: "#2da944",
                    version,
                    icons: {
                        android: true,
                        appleIcon: true,
                        appleStartup: true,
                        coast: true,
                        favicons: true,
                        firefox: false,
                        opengraph: true,
                        twitter: true,
                        yandex: false,
                        windows: true
                    }
                }
            }),
            new FileManagerPlugin({
                onStart: [
                    {
                        delete: ["./dist"]
                    }
                ],
                onEnd: [
                    {
                        copy: [
                            {
                                source: `./src/img/${brand}/resources/typography/*.png`,
                                destination: `./dist${basename}img/typography`
                            },
                            {
                                source: `./src/img/${brand}/resources/logotype/*.png`,
                                destination: `./dist${basename}img/logotype`
                            },
                            {
                                source: `./src/img/${brand}/documentation/buttons/*.png`,
                                destination: `./dist${basename}img/documentation/buttons`
                            },
                            {
                                source: `./src/img/${brand}/documentation/logotype/*.png`,
                                destination: `./dist${basename}img/documentation/logotype`
                            },
                            {
                                source: `./src/img/${brand}/documentation/cards/*.png`,
                                destination: `./dist${basename}img/documentation/cards`
                            },
                            {
                                source: `./src/img/${brand}/documentation/accessibility/*.png`,
                                destination: `./dist${basename}img/documentation/accessibility`
                            },
                            {
                                source: `./src/img/${brand}/documentation/imagery/*.svg`,
                                destination: `./dist${basename}img/documentation/imagery`
                            },
                            {
                                source: `./src/img/${brand}/documentation/spacing/*.svg`,
                                destination: `./dist${basename}img/documentation/spacing`
                            },
                            {
                                source: `./src/img/${brand}/documentation/copywriting/*.png`,
                                destination: `./dist${basename}img/documentation/copywriting`
                            },
                            {
                                source: `./src/img/${brand}/documentation/grid/*.png`,
                                destination: `./dist${basename}img/documentation/grid`
                            },
                            {
                                source: `./src/img/${brand}/documentation/colors/*.png`,
                                destination: `./dist${basename}img/documentation/colors`
                            },
                            {
                                source: `./src/img/${brand}/documentation/introduction/*.png`,
                                destination: `./dist${basename}img/documentation/introduction`
                            },
                            {
                                source: `./src/img/${brand}/documentation/forDevelopers/*.png`,
                                destination: `./dist${basename}img/documentation/forDevelopers`
                            },
                            {
                                source: `./src/img/${brand}/documentation/forDesigners/*.png`,
                                destination: `./dist${basename}img/documentation/forDesigners`
                            },
                            {
                                source: "./src/img/background/*.svg",
                                destination: `./dist${basename}img/background`
                            },
                            {
                                source: "./src/assets/logos/*.zip",
                                destination: `./dist${basename}release/logos`
                            },
                            {
                                source: "./src/assets/fonts/*.zip",
                                destination: `./dist${basename}release/fonts`
                            },
                            {
                                source: "./src/assets/templates/*",
                                destination: `./dist${basename}templates`
                            },
                            {
                                source: `./dist${basename}icons`,
                                destination: "./dist/temp/icons/icons"
                            },
                            {
                                source: `./dist${basename}scripts/dg.js`,
                                destination: "./dist/temp/release/scripts"
                            },
                            {
                                source: `./dist${basename}scripts/dg.js.map`,
                                destination: "./dist/temp/release/scripts"
                            },
                            {
                                source: `./dist${basename}scripts/dg-dashboard.js`,
                                destination: "./dist/temp/release/scripts"
                            },
                            {
                                source: `./dist${basename}scripts/dg-dashboard.js.map`,
                                destination: "./dist/temp/release/scripts"
                            },
                            {
                                source: `./dist${basename}styles/dg-style.css`,
                                destination: "./dist/temp/release/styles"
                            }
                        ],
                        mkdir: [`./dist${basename}release`],
                        archive: onEndArchive,
                        delete: ["./dist/temp"]
                    },
                    {
                        copy: appRoutes.map(route => ({
                            source: `./dist${basename}index.html`,
                            destination: `./dist${basename}${route}`
                        }))
                    }
                ]
            })
        );
    }

    return config;
};
